File: Leetcode/538-Convert-BST-To-Greater-Tree/ConvertBSTToGreaterTree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def convertBST(self, root):
        """
        :type root: TreeNode
        :rtype: TreeNode
        """
        if not root:
            return root
        self.convert(root, 0)
        return root

    def convert(self, node, sum):
        if not node:
            return sum
        node.val += self.convert(node.right, sum)
        return self.convert(node.left, node.val)
    

File: Leetcode/538-Convert-BST-To-Greater-Tree/ConvertBSTToGreaterTree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode* convertBST(TreeNode* root) {
        if (root == NULL) {
            return root;
        }
        convertBST(root, 0);
        return root;
    }
private :
    int convertBST(TreeNode* node, int sum) {
        if (node == NULL) {
            return sum;
        }
        node->val += convertBST(node->right, sum);
        return convertBST(node->left, node->val);
    }
};
